server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -669,7 +674,7 @@ private SearchResponse buildSearchResponse(
             getNumShards(),
             numSuccess,
             skippedOps.get(),
-            buildTookInMillis(),
+            searchResponseTookMetrics.record(buildTookInMillis()),
```

Comment:
You are now tracking all places where we provide a meaningful tookInMillis and adding the call to record the metric in those cases. That is fine, but I wonder if we can take it one step further. Can we introduce perhaps a new constructor argument to be able to better distinguish between places that do need tracking from places that do not? We could perhaps even add a static factory method that makes this easier to follow compared to a constructor with many arguments. Could we even make it so that the record call is done in one place, when creating that SearchResponse instance?

Reply:
If I understand correctly, you're suggesting passing the metric to SearchResponse so that it is responsible for calling the metric on construction? 

Hmm, I was originally thinking along those lines too, but decided that SearchResponse is a data container, and shouldn't really contain "behaviour", as this is mixing concerns. However, everything's a trade-off, and I agree it would add clarity.  I like your idea of using as a static factory, that feels less-bad while still resulting in the aforementioned benefits. I'll give it a whirl.

Reply:
going back to this, the logic to record could be outside of SearchResponse construction, in the static methods that you added.